src/kubernetes.cc

```diff
@@ -568,70 +569,111 @@ json::value KubernetesReader::QueryMaster(const std::string& path) const
 
 namespace {
 struct Watcher {
-  Watcher(std::function<void(json::value)> event_callback,
+  Watcher(const std::string& endpoint,
+          std::function<void(json::value)> event_callback,
           std::unique_lock<std::mutex>&& completion, bool verbose)
-      : completion_(std::move(completion)), event_callback_(event_callback),
-        remaining_chunk_bytes_(0), verbose_(verbose) {}
+      : name_("Watcher(" + endpoint + ")"),
+        completion_(std::move(completion)), event_callback_(event_callback),
+        remaining_chunk_bytes_(0), verbose_(verbose), exception_() {}
   ~Watcher() {}  // Unlocks the completion_ lock.
+
+ private:
+  // A representation of all watch-related errors.
+  class WatcherException {
+   public:
+    WatcherException(const std::string& what) : explanation_(what) {}
+    const std::string& what() const { return explanation_; }
+   private:
+    std::string explanation_;
+  };
+
+ public:
   void operator()(const boost::iterator_range<const char*>& range,
                   const boost::system::error_code& error) {
     if (!error) {
+      if (!exception_.empty()) {
```

Comment:
Being that exception is a string, can we change the name?

Reply:
Something like exception_message_ or exception_string_ just to add clarify that we can't treat it like an exception (despite it clearly being defined as a string)

Reply:
Renamed to `exception_message_`.